lib/core/heartbeat.cpp

```diff
@@ -60,9 +60,10 @@ namespace pstore {
     heartbeat::worker_thread::duration_type const heartbeat::worker_thread::max_time_ =
         duration_type::max ();
 
-    heartbeat::worker_thread::worker_thread ()
+    heartbeat::worker_thread::worker_thread (bool unittest)
             : sleep_time_ (&max_time_) {
-        threads::set_name ("heartbeat");
+        if (!unittest)
+            threads::set_name ("heartbeat");
```

Comment:
After the change, the unittest will not change the name of the thread. It looks good to me. Since Paul wrote the code and created the bug, Paul, could you please review it?

Reply:
I'm not a huge fan of this approach. Rather than embedding unit-test-ness into the ctor, could this instead be moved to heartbeat::attach(). We could have a version of attach() which takes an additional template function argument which is called from the lambda passed to std::thread.

Something like:

    template <typename SetupFunction>
    void heartbeat::attach (key_type key, callback cb, SetupFunction setup) {
        if (!state_) {
            state_ = std::make_unique<state> ();
            auto & w = state_->worker;
            state_->thread = std::thread ([&w, setup]() { setup(); w.run (); });
        }
        state_->worker.attach (key, cb);
    }

    void heartbeat::attach (key_type key, callback cb) {
        this->attach (key, cb, [] () {  threads::set_name ("heartbeat") });
    }

This way the unit tests can use the 3-argument version of attach().

Reply:
Comments added.

Reply:
My reasoning is that this moves the the conditional behaviour to compile time.

Reply:
I'm a little confused. I thought the problem was that the 'testHeartbeat' unit tests only test the worker thread class attach and detach. Thus when it changes the thread name it is confusing. 

If we change the unit test to call the heartbeat class attach and detach it will create a thread and changing the name will not matter? 

Do you want the testHeartbeat tests changed to call the heartbeat class?

Reply:
Good point. The problem is that these tests end up renaming the main thread. Creating threads in a unit test is best avoided where possible (debugging gets harder and theres a risk of unpredictable timings introducing random failures).

My concern is that I'm not keen on having the unit test behaviour conditional at runtime. If this can be done at compile-time using the technique I showed then that would be create.

I'm not asking for the testHeartbeat tests to call the heartbeat class.

Reply:
Thanks Paul. I'll re-jig the solution to control the thread renaming at compile time but limited to the worker class and not create any threads.